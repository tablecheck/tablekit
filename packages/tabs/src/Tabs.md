The `Tabs` component builds a scrollable navigation to display content divided in sections

### Examples

```tsx
import { useState } from 'react';
import styled from '@emotion/styled';
import {
  DARK_COLORS,
  CLASSIC_COLORS,
  ThemeProvider
} from '@tablecheck/tablekit-theme';
import { Checkbox } from '@tablecheck/tablekit-checkbox';
import { Header } from '@tablecheck/tablekit-typography';
import {
  Badge,
  badgeThemeNamespace,
  badgeDarkTheme,
  badgeClassicTheme
} from '@tablecheck/tablekit-badge';
import { Icon } from '@tablecheck/tablekit-icon';
import { faOmega } from '@fortawesome/pro-solid-svg-icons/faOmega';
import { ThemeProvider as EmotionThemeProvider } from '@emotion/react';
import { tabsDarkTheme, tabsClassicTheme, tabsThemeNamespace } from './index';

const TabContent = styled.div`
  display: flex;
  width: 100%;
`;

const Content = styled.div`
  padding: 20px;
  ${({ isActive }) => (isActive ? 'display: block;' : 'display: none;')};
  ${({ theme }) => `color: ${theme.colors.text};`}
`;

const ItemBadge = styled.div`
  display: inline-flex;
  align-items: center;
`;

const ItemIcon = styled(Icon)`
  margin-left: 5px;
`;

const ExamplesWrapper = styled.div`
  ${({ theme }) => `background-color: ${theme.colors.canvas};`}
  padding: 10px;
`;

const StyledBadge = styled(Badge)`
  margin-left: 5px;
`;

const tabs = [
  {
    content: <span>Tab 1</span>,
    key: 'tab-1'
  },
  {
    content: <span>Tab 2</span>,
    key: 'tab-2',
    isDisabled: true
  },
  {
    content: <span>Tab 3</span>,
    key: 'tab-3'
  },
  {
    content: <span>Tab 4</span>,
    key: 'tab-4'
  }
];

const tabs2 = [
  {
    content: (
      <ItemBadge>
        <span>Shops</span>
        <StyledBadge size="small" value={2} isFlexible />
      </ItemBadge>
    ),
    key: 'tab-1'
  },
  {
    content: (
      <ItemBadge>
        <span>Customers</span>
        <ItemIcon icon={faOmega} />
      </ItemBadge>
    ),
    key: 'tab-2'
  },
  {
    content: (
      <ItemBadge>
        <span>Reservations</span>
        <StyledBadge size="small" value={44} isFlexible />
      </ItemBadge>
    ),
    key: 'tab-3'
  },
  {
    content: <span>Tables</span>,
    key: 'tab-4'
  }
];

const PresentationModule = () => {
  const [isDarkMode, setDarkMode] = useState(false);
  const [currentTab, setCurrentTab] = useState('tab-1');
  const [currentTab2, setCurrentTab2] = useState('tab-1');
  const [currentTab3, setCurrentTab3] = useState('tab-3');

  const darkTheme = {
    colors: DARK_COLORS,
    [tabsThemeNamespace]: tabsDarkTheme,
    [badgeThemeNamespace]: badgeDarkTheme
  };

  const classicTheme = {
    colors: CLASSIC_COLORS,
    [tabsThemeNamespace]: tabsClassicTheme,
    [badgeThemeNamespace]: badgeClassicTheme
  };

  return (
    <div>
      <Checkbox onChange={() => setDarkMode(!isDarkMode)}>
        Use dark theme
      </Checkbox>
      <EmotionThemeProvider theme={isDarkMode ? darkTheme : classicTheme}>
        <ExamplesWrapper>
          <div>
            <Tabs
              tabs={tabs}
              currentTab={currentTab}
              setCurrentTab={setCurrentTab}
            />
            <TabContent>
              <Content isActive={currentTab === 'tab-1'}>
                LoAenean et dapibus massa. Cras sagittis maximus urna. In
                egestas, libero vel egestas ornare, purus urna faucibus urna, et
                accumsan nisi velit vitae arcu. Maecenas tempor orci scelerisque
                sapien tristique semper. Vivamus scelerisque metus ipsum, in
                sagittis tortor blandit ac. Mauris sit amet metus non erat
                mollis suscipit quis eu libero.
              </Content>
              <Content isActive={currentTab === 'tab-2'}>
                Convallis ac ligula. Curabitur ut neque eget velit mollis
                feugiat vitae sit amet nibh. Integer quis accumsan nulla. In
                laoreet bibendum volutpat. Aliquam molestie sapien sit amet eros
                luctus, sit amet laoreet eros pretium. Aenean et dapibus massa.
                Cras sagittis maximus urna. In egestas, libero vel egestas
                ornare, purus urna faucibus urna, et accumsan nisi velit vitae
                arcu. Maecenas tempor orci scelerisque sapien tristique semper.
                Vivamus scelerisque metus ipsum, in sagittis tortor blandit ac.
                Mauris sit amet metus non erat mollis suscipit quis eu libero.
              </Content>
              <Content isActive={currentTab === 'tab-3'}>
                Aliquam molestie sapien sit amet eros luctus, sit amet laoreet
                eros pretium. Aenean et dapibus massa. Cras sagittis maximus
                urna. In egestas, libero vel egestas ornare, purus urna faucibus
                urna, et accumsan nisi velit vitae arcu. Maecenas tempor orci
                scelerisque sapien tristique semper. Vivamus scelerisque metus
                ipsum, in sagittis tortor blandit ac. Mauris sit amet metus non
                erat mollis suscipit quis eu libero.
              </Content>
              <Content isActive={currentTab === 'tab-4'}>
                Lorem ipsum dolor sit amet, consectetur adipiscing elit. Proin
                sed porta lorem. Proin laoreet mauris sit amet pretium suscipit.
                Etiam ut gravida justo. Vivamus placerat maximus nulla.
                Vestibulum odio enim, condimentum maximus auctor nec, convallis
                ac ligula. Curabitur ut neque eget velit mollis feugiat vitae
                sit amet nibh. Integer quis accumsan nulla. In laoreet bibendum
                volutpat. libero.
              </Content>
            </TabContent>
            <Tabs
              align="center"
              tabs={tabs2}
              currentTab={currentTab2}
              setCurrentTab={setCurrentTab2}
            />
            <TabContent>
              <Content isActive={currentTab2 === 'tab-1'}>One</Content>
              <Content isActive={currentTab2 === 'tab-2'}>Two</Content>
              <Content isActive={currentTab2 === 'tab-3'}>Three</Content>
              <Content isActive={currentTab2 === 'tab-4'}>Four</Content>
            </TabContent>
            <Tabs
              align="right"
              tabs={tabs}
              currentTab={currentTab3}
              setCurrentTab={setCurrentTab3}
            />
            <TabContent>
              <Content isActive={currentTab3 === 'tab-1'}>One</Content>
              <Content isActive={currentTab3 === 'tab-2'}>Two</Content>
              <Content isActive={currentTab3 === 'tab-3'}>Three</Content>
              <Content isActive={currentTab3 === 'tab-4'}>Four</Content>
            </TabContent>
          </div>
        </ExamplesWrapper>
      </EmotionThemeProvider>
    </div>
  );
};

<ThemeProvider useLocalStyles>
  <PresentationModule />
</ThemeProvider>;
```
