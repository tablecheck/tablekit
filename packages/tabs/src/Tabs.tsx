import {
  IS_CLICKED_ATTR,
  NavLine,
  TabItem,
  TabsNavigation,
  TabsWrapper
} from './styled';
import { TabsAlignment, TabsProps } from './types';

/*
 * Note about the usage of JS "classList" in this component.
 *
 * As this component should be essentially stateless and we can handle all the
 * styling via the pseudo-styles of :hover :active etc there is one edge case
 * that requires some extra javascript work. This case is in the styling of :focus.
 * In the browser :focus will be active if you focus via keyboard or mouse/touch.
 * There is a proposal to add :focus-visible that will replace this logic.
 *
 * So to avoid both the extra renders from state updates and the extra work
 * @emotion would need to change the stylesheets+classes from these state updates
 * we are simply using a basic class restrict when :focus styling is applied.
 *
 * As tab buttons can quite often be used in very large amounts this is a
 * reasonable trade off against being a complete React way.
 */
let isKeyboardInteraction = false;

export const Tabs = ({
  className,
  align = TabsAlignment.Left,
  tabs,
  currentTab,
  setCurrentTab,
  ...restProps
}: TabsProps): JSX.Element => (
  <TabsWrapper {...restProps} className={className}>
    <TabsNavigation align={align}>
      {tabs.map((tab) => {
        const isActive = currentTab === tab.key;
        return (
          <TabItem
            aria-selected={isActive}
            data-value={tab.key}
            id={tab.id}
            name={tab.name}
            key={tab.key}
            isActive={isActive}
            disabled={tab.isDisabled}
            onClick={(event) => {
              if (isKeyboardInteraction) return;
              event.currentTarget.dataset[IS_CLICKED_ATTR] = 'true';
              if (!tab.isDisabled) setCurrentTab(tab.key);
            }}
            onKeyDown={() => {
              isKeyboardInteraction = true;
            }}
            onMouseDown={(event) => {
              event.currentTarget.dataset[IS_CLICKED_ATTR] = 'true';
              isKeyboardInteraction = false;
            }}
            onBlur={(event) => {
              delete event.currentTarget.dataset[IS_CLICKED_ATTR];
            }}
          >
            {tab.content}
            {isActive && <NavLine />}
          </TabItem>
        );
      })}
    </TabsNavigation>
  </TabsWrapper>
);
