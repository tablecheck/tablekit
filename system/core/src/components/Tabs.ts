import { css } from '@emotion/react';

import { OptionalKeys } from '../typeUtils';

export const className = 'tabs';
export interface Props {
  role: 'tablist';
}

export type DefaultedProps = OptionalKeys<Props, 'role'>;
export const defaultProps = { role: 'tablist' };

export const baseStyles = css`
  display: flex;
  flex-direction: row;
  gap: var(--spacing-l5);
`;
