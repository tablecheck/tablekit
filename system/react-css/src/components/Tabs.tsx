/**
 * DO NOT EDIT: This file is generated, run 'npm update:components' to update this.
 * The exports here are generated from @tablecheck/tablekit-core
 * If you need to provide more "structure" to this component move it to the 'structuredComponents' folder
 */
import type { tabs } from '@tablecheck/tablekit-core';
import * as React from 'react';

export type Props = tabs.DefaultedProps;

export const Tabs = React.forwardRef<
  HTMLDivElement,
  Props & React.HTMLAttributes<HTMLDivElement>
>((props, ref) => (
  <div
    role="tablist"
    {...props}
    ref={ref}
    className={`${props.className || ''} tabs`}
  />
));
