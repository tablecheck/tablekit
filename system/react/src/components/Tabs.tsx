/**
 * DO NOT EDIT: This file is generated, run 'npm update:components' to update this.
 * The exports here are generated from @tablecheck/tablekit-core
 * If you need to provide more "structure" to this component move it to the 'structuredComponents' folder
 */
import styled from '@emotion/styled';
import { tabs } from '@tablecheck/tablekit-core';

export type Props = tabs.DefaultedProps;

export const Tabs = styled.div<Props>`
  ${tabs.baseStyles}
`;
Tabs.defaultProps = { role: 'tablist' };
