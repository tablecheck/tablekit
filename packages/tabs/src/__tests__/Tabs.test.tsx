import { mount } from '@tablecheck/tablekit-enzyme';

import { Tabs } from '../Tabs';

beforeEach(() => {
  global.console.error = jest.fn();
});

describe('Tabs', () => {
  test('should be possible to create a component', () => {
    const wrapper = mount(
      <Tabs currentTab="" setCurrentTab={() => {}} tabs={[]} />
    );
    expect(wrapper).not.toBeUndefined();
  });

  test('should correctly apply styling', () => {
    expect(
      mount(<Tabs currentTab="" setCurrentTab={() => {}} tabs={[]} />)
    ).toMatchSnapshot();
  });

  test('should allow passing of custom className', () => {
    expect(
      mount(
        <Tabs
          currentTab=""
          setCurrentTab={() => {}}
          tabs={[]}
          className="test-class"
        />
      )
    ).toMatchSnapshot();
  });
});
