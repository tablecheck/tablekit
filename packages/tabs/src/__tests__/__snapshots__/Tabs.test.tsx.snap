// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs should allow passing of custom className 1`] = `
.emotion-0 {
  position: relative;
}

.emotion-2 {
  width: 100%;
  overflow-x: auto;
  white-space: nowrap;
  font-size: 14px;
  line-height: 20px;
  font-weight: 400;
  text-align: left;
}

.emotion-2:after {
  content: '';
  position: relative;
  display: block;
  border-bottom: 1px solid #E4E4E4;
}

<Tabs
  className="test-class"
  currentTab=""
  setCurrentTab={[Function]}
  tabs={Array []}
>
  <TabsWrapper
    className="test-class"
  >
    <div
      className="test-class emotion-0 emotion-1"
    >
      <TabsNavigation
        align="left"
      >
        <div
          className="emotion-2 emotion-3"
        />
      </TabsNavigation>
    </div>
  </TabsWrapper>
</Tabs>
`;

exports[`Tabs should correctly apply styling 1`] = `
.emotion-0 {
  position: relative;
}

.emotion-2 {
  width: 100%;
  overflow-x: auto;
  white-space: nowrap;
  font-size: 14px;
  line-height: 20px;
  font-weight: 400;
  text-align: left;
}

.emotion-2:after {
  content: '';
  position: relative;
  display: block;
  border-bottom: 1px solid #E4E4E4;
}

<Tabs
  currentTab=""
  setCurrentTab={[Function]}
  tabs={Array []}
>
  <TabsWrapper>
    <div
      className="emotion-0 emotion-1"
    >
      <TabsNavigation
        align="left"
      >
        <div
          className="emotion-2 emotion-3"
        />
      </TabsNavigation>
    </div>
  </TabsWrapper>
</Tabs>
`;
